### ISSUE TITLE
How to obtain extrinsic matrix given cam_position and inplane_rotation?

### ISSUE THREAD

**Issue by xingyi-li** [xingyuansun/pix3d#25]
Hi, Xingyuan!

I would like to test models on Pix3D, and I want to obtain the extrinsic matrix where the origin of the world coordinate system is at the center of the object and the camera is looking at the center of object. I notice that "rot_mat" and "trans_mat" are not the "Rt" I want.

Therefore, I would like to ask that how I can obtain extrinsic matrix given cam_position and inplane_rotation.

Best,
Xingyi

**Comment by xingyuansun** [comment]
Hi Xingyi,

Thank you for using Pix3D! Given `cam_position` and `inplane_rotation`, the translation is simply `cam_position`. To obtain the rotation matrix, first perform two rotations on the yaw axis and the pitch axis, whose angles can be calculated from `cam_position`. We then perform a rotation on the roll axis, whose angle is `inplane_rotation`. You may find more about yaw, pitch, roll from [here](https://en.wikipedia.org/wiki/Aircraft_principal_axes). Thank you!

Best,
Xingyuan

**Comment by xingyi-li** [comment]
Hi Xingyuan,

I can understand what you mean, and I write the following code according to your words. However, I still fail to get a normal result. Would you please help me check out if I am correct? By the way, after calling ```get_extrinsic``` and concatenating with ```cam_position```, I get c2w matrix (am I right?), and the coordinate system is +x: leftward, +y: upward, +z: inward, right?

```python
def rot_pitch(pitch):
    return torch.tensor(
        [
            [1, 0, 0],
            [0, np.cos(pitch), -np.sin(pitch)],
            [0, np.sin(pitch), np.cos(pitch)],
        ],
        dtype=torch.float32,
    )


def rot_yaw(yaw):
    return torch.tensor(
        [
            [np.cos(yaw), 0, np.sin(yaw)],
            [0, 1, 0],
            [-np.sin(yaw), 0, np.cos(yaw)],
        ],
        dtype=torch.float32,
    )


def rot_roll(roll):
    return torch.tensor(
        [
            [np.cos(roll), -np.sin(roll), 0],
            [np.sin(roll), np.cos(roll), 0],
            [0, 0, 1],
        ],
        dtype=torch.float32,
    )


def angles_to_matrix(angles):
    azi = angles[:, 0]
    ele = angles[:, 1]
    rol = angles[:, 2]

    rot_mat = rot_roll(rol) @ rot_pitch(ele) @ rot_yaw(azi)
    return rot_mat

def get_extrinsic(pix3d_info):
    cam_position = pix3d_info['cam_position']
    elevation = math.degrees(math.atan(cam_position[1] / math.sqrt(cam_position[0]**2 + cam_position[2]**2)))
    if cam_position[2] == 0. or cam_position[2] == -0.:
        if cam_position[0] <= 0:
            azimuth = 90.
        else:
            azimuth = 270.
    else:
        azimuth = math.degrees(math.atan(cam_position[0] / cam_position[2]))

    inplane_rotation = pix3d_info['inplane_rotation']
    azimuth = math.radians(azimuth)
    elevation = math.radians(elevation)
    angles = torch.tensor([azimuth, elevation, inplane_rotation]).reshape(1, -1)
    rot_mat = angles_to_matrix(angles)

    return rot_mat
```

Best,
Xingyi

**Comment by xingyuansun** [comment]
Hi Xingyi,

Thank you for your reply! The code looks correct to me. Note that `cam_position` and `inplane_rotation` are an approximation to `rot_mat` and `trans_mat` we provided, since the camera is not always looking at the object's center. Thanks!

Best,
Xingyuan
